### PR TITLE
Doc: references for `lfilter` implementation

### PR DESCRIPTION
**Pull request by yoyololicon**
Add references to @yoyololicon and @forgi86's works for the core implementation of `lfilter`.
@forgi86 implemented the same logic as torchaudio's in [forgi86/dynonet](https://github.com/forgi86/dynonet) for efficient backpropagation.

### CODE REVIEW

File: src/torchaudio/functional/filtering.py [hunk 1]
```diff
@@ -998,7 +998,7 @@ def _lfilter_core(
 
 
 def lfilter(waveform: Tensor, a_coeffs: Tensor, b_coeffs: Tensor, clamp: bool = True, batching: bool = True) -> Tensor:
-    r"""Perform an IIR filter by evaluating difference equation.
+    r"""Perform an IIR filter by evaluating difference equation, using differentiable implementation developed independently by *Yu et al.* :cite:`ismir_YuF23` and *Forgione et al.* :cite:`forgione2021dynonet`.
```
Comment (mthrok) [inline src/torchaudio/functional/filtering.py:1001]: `flake8` is freaking out. 

https://github.com/pytorch/audio/actions/runs/7630525757/job/20786415251?pr=3734

Could you make this into a separate line?

**Review by mthrok — APPROVED** [review]
Thanks for the PR. Looks good but the linter is failing.

File: src/torchaudio/functional/filtering.py (hunk 1 above)
Comment (yoyololicon) [reply src/torchaudio/functional/filtering.py]: Fixed.

File: src/torchaudio/functional/filtering.py [hunk 2]
```diff
@@ -998,7 +998,8 @@ def _lfilter_core(
 
 
 def lfilter(waveform: Tensor, a_coeffs: Tensor, b_coeffs: Tensor, clamp: bool = True, batching: bool = True) -> Tensor:
-    r"""Perform an IIR filter by evaluating difference equation.
+    r"""Perform an IIR filter by evaluating difference equation, using differentiable implementation
+        developed independently by *Yu et al.* :cite:`ismir_YuF23` and *Forgione et al.* :cite:`forgione2021dynonet`.
```
Comment (mthrok) [inline src/torchaudio/functional/filtering.py:1002]:
```suggestion
    developed independently by *Yu et al.* :cite:`ismir_YuF23` and *Forgione et al.* :cite:`forgione2021dynonet`.
```

I think the extra indentation is causing the first line to become bold for some reason. 

## Current Doc
https://pytorch.org/audio/2.2.0/generated/torchaudio.functional.lfilter.html#torchaudio.functional.lfilter

Screenshot 2024-01-25 at 3 33 14 PM

## Current PR state
https://docs-preview.pytorch.org/pytorch/audio/3734/generated/torchaudio.functional.lfilter.html#torchaudio.functional.lfilter

Screenshot 2024-01-25 at 3 33 19 PM

Comment (yoyololicon) [reply src/torchaudio/functional/filtering.py]: Thanks for the catch. Sorry my local computer can't build the doc for some reason so I can't preview the doc immediately. 😅

Comment (yoyololicon) [reply src/torchaudio/functional/filtering.py]: It should work now.
image

**Review by mthrok — APPROVED** [review]
LGTM

**Comment by mthrok** [comment]
Feel free to merge!

**Pull request merged**